libraries/AP_Camera/AP_Camera.cpp

```diff
@@ -244,69 +228,70 @@ bool AP_Camera::control(float session, float zoom_pos, float zoom_step, float fo
+void AP_Camera::send_feedback(mavlink_channel_t chan)
 {
     float altitude, altitude_rel;
-    if (current_loc.flags.relative_alt) {
-        altitude = current_loc.alt+ahrs.get_home().alt;
-        altitude_rel = current_loc.alt;
+    if (current_loc->flags.relative_alt) {
+        altitude = current_loc->alt+ahrs->get_home().alt;
+        altitude_rel = current_loc->alt;
     } else {
-        altitude = current_loc.alt;
-        altitude_rel = current_loc.alt - ahrs.get_home().alt;
+        altitude = current_loc->alt;
+        altitude_rel = current_loc->alt - ahrs->get_home().alt;
     }
 
     mavlink_msg_camera_feedback_send(chan, 
-        gps.time_epoch_usec(),
+        gps->time_epoch_usec(),
         0, 0, _image_index,
-        current_loc.lat, current_loc.lng,
+        current_loc->lat, current_loc->lng,
         altitude/100.0f, altitude_rel/100.0f,
-        ahrs.roll_sensor/100.0f, ahrs.pitch_sensor/100.0f, ahrs.yaw_sensor/100.0f,
+        ahrs->roll_sensor/100.0f, ahrs->pitch_sensor/100.0f, ahrs->yaw_sensor/100.0f,
         0.0f,CAMERA_FEEDBACK_PHOTO);
 }
 
 
-/*  update location, for triggering by GPS distance moved
-    This function returns true if a picture should be taken
-    The caller is responsible for taking the picture based on the return value of this function.
-    The caller is also responsible for logging the details about the photo
+/*  update; triggers by distance moved
 */
-bool AP_Camera::update_location(const struct Location &loc, const AP_AHRS &ahrs)
+void AP_Camera::update()
 {
+    if (gps->status() < AP_GPS::GPS_OK_FIX_3D) {
+        return;
+    }
+
     if (is_zero(_trigg_dist)) {
-        return false;
+        return;
     }
     if (_last_location.lat == 0 && _last_location.lng == 0) {
-        _last_location = loc;
-        return false;
+        _last_location = *current_loc;
+        return;
     }
-    if (_last_location.lat == loc.lat && _last_location.lng == loc.lng) {
-        // we haven't moved - this can happen as update_location() may
+    if (_last_location.lat == current_loc->lat && _last_location.lng == current_loc->lng) {
+        // we haven't moved - this can happen as update() may
         // be called without a new GPS fix
-        return false;
+        return;
     }
 
-    if (get_distance(loc, _last_location) < _trigg_dist) {
-        return false;
+    if (get_distance(*current_loc, _last_location) < _trigg_dist) {
+        return;
     }
 
-    if (_max_roll > 0 && labs(ahrs.roll_sensor/100) > _max_roll) {
-        return false;
+    if (_max_roll > 0 && labs(ahrs->roll_sensor/100) > _max_roll) {
+        return;
     }
 
     uint32_t tnow = AP_HAL::millis();
     if (tnow - _last_photo_time < (unsigned) _min_interval) {
-        return false;
-    }  else {
-        _last_location = loc;
-        _last_photo_time = tnow;
-        return true;
+        return;
     }
+
+    take_picture();
+
+    _last_location = *current_loc;
```

Comment:
Disagree.  Someone may choose to do a trigger just-because - that shouldn't go out of its way to screw up the distance-based grid-pattern thing that the vehicle may be trying to fly.

In fact, possibly we should reject taking random pictures while doing the distance-based-triggering.